packages/salesforcedx-vscode-core/src/commands/baseDeployRetrieve.ts

```diff
@@ -0,0 +1,390 @@
+  protected abstract postOperation(
+    result: DeployRetrieveResult | undefined
+  ): Promise<void>;
+}
+
+export abstract class DeployCommand<T> extends DeployRetrieveExecutor<T> {
+  protected async doOperation(
+    components: ComponentSet
+  ): Promise<DeployResult | undefined> {
+    return components
+      .deploy({
+        usernameOrConnection: await workspaceContext.getConnection()
+      })
+      .start();
+  }
+
+  protected async postOperation(
+    result: DeployResult | undefined
+  ): Promise<void> {
+    try {
+      if (result) {
+        BaseDeployExecutor.errorCollection.clear();
+
+        const relativePackageDirs = await SfdxPackageDirectories.getPackageDirectoryPaths();
+        const output = this.createOutput(result, relativePackageDirs);
+        channelService.appendLine(output);
+
+        const success = result.response.status === RequestStatus.Succeeded;
+
+        if (!success) {
+          handleDeployDiagnostics(result, BaseDeployExecutor.errorCollection);
+        }
+      }
+    } finally {
+      await DeployQueue.get().unlock();
+    }
+  }
+
+  private createOutput(
+    result: DeployResult,
+    relativePackageDirs: string[]
+  ): string {
+    const table = new Table();
+
+    const rowsWithRelativePaths = (result.getFileResponses().map(response => {
+      response.filePath = this.getRelativeProjectPath(
+        response.filePath,
+        relativePackageDirs
+      );
+      return response;
+    }) as unknown) as Row[];
+
+    let output: string;
+
+    if (result.response.status === RequestStatus.Succeeded) {
+      output = table.createTable(
+        rowsWithRelativePaths,
+        [
+          { key: 'state', label: nls.localize('table_header_state') },
+          { key: 'fullName', label: nls.localize('table_header_full_name') },
+          { key: 'type', label: nls.localize('table_header_type') },
+          {
+            key: 'filePath',
+            label: nls.localize('table_header_project_path')
+          }
+        ],
+        nls.localize(`table_title_deployed_source`)
+      );
+    } else {
+      output = table.createTable(
+        rowsWithRelativePaths.filter(row => row.error),
+        [
+          {
+            key: 'filePath',
+            label: nls.localize('table_header_project_path')
+          },
+          { key: 'error', label: nls.localize('table_header_errors') }
+        ],
+        nls.localize(`table_title_deploy_errors`)
+      );
+    }
+
+    return output;
+  }
+}
+
+export abstract class RetrieveCommand<T> extends DeployRetrieveExecutor<T> {
+  protected async doOperation(
+    components: ComponentSet
+  ): Promise<RetrieveResult | undefined> {
+    const connection = await workspaceContext.getConnection();
+
+    // utilize the tooling API for single component retrieves for improved performance
+    const oneComponent = components.getSourceComponents().first();
+
+    if (components.size === 1 && this.isToolingSupported(oneComponent)) {
```

Comment:
All retrieve commands will utilize the tooling api when it can, since single component retrieves are quite fast.

Reply:
Why are we not adding this into SDR ?

Reply:
This is how it has always been in the extensions up until this point. It should be added to the library though.